File: demos/nasm/hexdump.c
/* Minimal `hexdump -C` for hosts without one (native Windows). NASM's
 * nasm-t.py shells out to `hexdump -C <file>` to render a byte diff when a
 * golden test fails; it is display-only and not part of the pass/fail verdict.
 * Built with badc so the demo needs no external tool. Reads the last argument
 * as the path (nasm-t.py invokes it as `hexdump -C <file>`). */
#include <stdio.h>

int main(int argc, char **argv) {
    if (argc < 2)
        return 1;
    FILE *f = fopen(argv[argc - 1], "rb");
    if (!f)
        return 1;
    unsigned char buf[16];
    unsigned long off = 0;
    size_t n;
    while ((n = fread(buf, 1, sizeof buf, f)) > 0) {
        printf("%08lx  ", off);
        for (size_t i = 0; i < 16; i++) {
            if (i < n)
                printf("%02x ", buf[i]);
            else
                printf("   ");
            if (i == 7)
                printf(" ");
        }
        printf(" |");
        for (size_t i = 0; i < n; i++) {
            unsigned char c = buf[i];
            putchar((c >= 32 && c < 127) ? c : '.');
        }
        printf("|\n");
        off += (unsigned long)n;
    }
    printf("%08lx\n", off);
    fclose(f);
    return 0;
}

File: demos/nasm/hexdump.py
#!/usr/bin/env python3
"""Minimal `hexdump -C` for hosts without one (native Windows).

NASM's nasm-t.py shells out to `hexdump -C <file>` to render a byte diff when a
golden test fails; it is display-only and does not affect the pass/fail
decision. This reproduces the canonical `-C` layout so the suite runs where the
BSD/util-linux hexdump is absent.
"""
import sys


def main() -> int:
    path = sys.argv[-1]
    with open(path, "rb") as f:
        data = f.read()
    for off in range(0, len(data), 16):
        chunk = data[off:off + 16]
        hi = " ".join(f"{b:02x}" for b in chunk[:8])
        lo = " ".join(f"{b:02x}" for b in chunk[8:])
        hexpart = f"{hi:<23}  {lo:<23}"
        ascii_part = "".join(chr(b) if 32 <= b < 127 else "." for b in chunk)
        print(f"{off:08x}  {hexpart}  |{ascii_part}|")
    print(f"{len(data):08x}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
